tg_bot/modules/users.py

```diff
@@ -65,6 +68,136 @@ def broadcast(bot: Bot, update: Update):
                                             "due to being kicked.".format(failed))
 
 
+@run_async
+def restrict_group(bot: Bot, update: Update, args: List[str]) -> str:
+    message = update.effective_message
+
+    # Check if there is only one argument
+    if len(args) == 1:
+        chat_id = args[0]
+
+        # Check if chat_id is valid
+        if chat_id.startswith('-'):
+            # Check if chat_id is in bot database
+            all_chats = sql.get_all_chats() or []
+            if any(chat.chat_id == chat_id for chat in all_chats):
+                chat_restricted = sql.get_restriction(chat_id)    
+                if not chat_restricted:
+                    try:
+                        chat_title = bot.get_chat(chat_id).title
+
+                        sudo_users_list = "<b>My Admins:</b>"
+                        for user in SUDO_USERS:
+                            name = "<a href=\"tg://user?id={}\">{}</a>".format(user, bot.get_chat(user).first_name)
+                            sudo_users_list += "\n - {}".format(name)
+
+                        bot.send_message(chat_id = chat_id,
+                                         text = "I have been restricted by my admins from this chat. "
+                                                "Request any of my admins to add me to this chat.\n\n"
+                                                "{admins_list}".format(admins_list = sudo_users_list),
+                                         parse_mode = ParseMode.HTML)
+
+                        bot.leave_chat(chat_id)
+
+                        sql.set_restriction(chat_id, restricted = True)
+
+                        message.reply_text("Successfully left chat <b>{}</b>!".format(chat_title),
+                                           parse_mode = ParseMode.HTML)
+            
+                        # Report to sudo users
+                        restrictor = update.effective_user  # type: Optional[User]
+                        send_to_list(bot, SUDO_USERS,
+                                     "{} has restricted me from being added to the chat <b>{}</b>."
+                                     .format(mention_html(restrictor.id, restrictor.first_name), chat_title),
+                                     html=True)
+
+                    except BadRequest as excp:
```

Comment:
This is a VERY large try/except block. make sure you're only wrapping the line which might throw the exception you're expecting.